addons/point_of_sale/models/pos_order.py

```diff
@@ -750,36 +831,37 @@ def action_pos_order_done(self):
         return self._create_account_move_line()
 
     @api.model
-    def create_from_ui(self, orders):
-        # Keep only new orders
-        submitted_references = [o['data']['name'] for o in orders]
-        pos_order = self.search([('pos_reference', 'in', submitted_references)])
-        existing_orders = pos_order.read(['pos_reference'])
-        existing_references = set([o['pos_reference'] for o in existing_orders])
-        orders_to_save = [o for o in orders if o['data']['name'] not in existing_references]
+    def create_from_ui(self, orders, draft):
+        """ Create and update Orders from the frontend PoS application.
+
+        Create new orders and update orders that are in draft status. If an order already exists with a status
+        diferent from 'draft'it will be discareded, otherwise it will be saved to the database. If saved with 
+        'draft' status the order can be overwritten later by this function.
+
+        :param orders: dictionary with the orders to be created.
+        :type orders: dict.
+        :param draft: Indicate if the orders are ment to be finalised or temporarily saved.
+        :type draft: bool.
+        :Returns: list -- list of db-ids for the created and updated orders.
+        """
+        orders_to_save = []
+        orders_to_update = []
         order_ids = []
 
-        for tmp_order in orders_to_save:
-            to_invoice = tmp_order['to_invoice']
-            order = tmp_order['data']
-            if to_invoice:
-                self._match_payment_to_invoice(order)
-            pos_order = self._process_order(order)
-            order_ids.append(pos_order.id)
+        existing_orders = self.search([('id', 'in', [o['data']['server_id'] for o in orders if 'server_id' in o['data']]),('state', '=', 'draft')])
+        orders_to_save = [o for o in orders if 'server_id' not in o['data'] or not o['data']['server_id']]
 
-            try:
-                pos_order.action_pos_order_paid()
-            except psycopg2.DatabaseError:
-                # do not hide transactional errors, the order(s) won't be saved!
-                raise
-            except Exception as e:
-                _logger.error('Could not fully process the POS Order: %s', tools.ustr(e))
+        orders_to_update = [o for o in orders if 'server_id' in o['data'] and o['data']['server_id'] in existing_orders.ids]
 
-            if to_invoice:
-                pos_order.action_pos_order_invoice()
-                pos_order.invoice_id.sudo().with_context(force_company=self.env.user.company_id.id).action_invoice_open()
-                pos_order.account_move = pos_order.invoice_id.move_id
-        return order_ids
+        for order in orders_to_save:
```

Comment:
you can put all this logic and the one in '_create_or_update_order' into '_process_order' it'll avoid to check multiple times if there are existing ordrers, and the method '_create_or_update_order' is only used here.

Reply:
still relevent